staging/src/k8s.io/client-go/tools/record/event.go

```diff
@@ -121,6 +121,10 @@ type EventBroadcaster interface {
 	// function. The return value can be ignored or used to stop recording, if desired.
 	StartLogging(logf func(format string, args ...interface{})) watch.Interface
 
+	// StartStructuredLogging starts sending events received from this EventBroadcaster to the structured
+	// logging function. The return value can be ignored or used to stop recording, if desired.
+	StartStructuredLogging(verbosity klog.Level) watch.Interface
```

Comment:
this breaks everybody implementing EventBroadcaster.

Reply:
Do we care about external to k/k code dependencies? Recently client-go changed it's interface to use context. 
I think if such change goes to minor (1.19) release it should be ok.

Reply:
We do.

Do the due diligence that this is not a common problem at least. Github code search might tell.

Reply:
I think there is no disagreement that we want event logs to support structured logging?
So what alternative solution (to growing interface) would you suggest?

I expect that never changing interface to avoid breaking change or reusing `StartLogging` for structured logging are not an option.

Reply:
/cc @lavalamp

Reply:
If this will break existing users (do a search) then we have to make a parallel interface type and do a runtime check to see if passed things support the parallel interface.

Reply:
Adding context warranted breaking people. This one doesn't. Sorry for failing to notice that the interface could be used that way earlier.

Reply:
@sttts are you aware of people implementing this interface? I wouldn't actually expect to find that, but I haven't searched.

Reply:
Not a specialist in GitHub search but I spent some time on trying to write a proper query. I tried many many different methods I have ended up with 
https://github.com/search?q=%22func%22+%22StartRecordingToSink%22+%22EventSink%22++-filename%3Aevent.go+language%3Ago+-filename%3Ainterfaces.go+-filename%3A*.test.go+NOT+%22EventSinkImpl%22+-filename%3Adynamic.go+NOT+%22NewFederatedEventSink%22&type=Code

I was looking for implementation of `StartRecordingToSink` which should look like `func (x *X) StartRecordingToSink(y EventSink) z.Interface`. GitHub doesn't allow pattern matching so I needed to look for `func`, `StartRecordingToSink`, `EventSink` as separate strings. I also added filters to remove some noise `-filename:event.go language:go -filename:interfaces.go -filename:*.test.go NOT "EventSinkImpl" -filename:dynamic.go NOT "NewFederatedEventSink"`. Noise is generates as there are hundreds of copies of Kubernetes code on GitHub (not forks, not vendoring just copies). 

To check correctness of this method, upon removal of `-filename:event.go` it should find a file with current implemenation. It finds also 400 copies of this implementation so I didn't check. It's possible that someone implemented custom version and put it in `event.go` file. It's hard to verify this without looking at every copy as I didn't find a way to distinquish them. I tried using path (K8s version is always in `client-go`), but I didn't manage to get `NOT client-go in:path` GitHub filter to work correctly (quering for both path and code didn't give correct results).

@sttts please let me know if this research makes sense and what could we do better to verify if noone reimplements EventBroadcaster. Even if method is not ideal I would expect to stumble upon at least one case. Would be good to know if we can move forward.

Reply:
I also did a quick search without find anything. I also checked our pretty big OpenShift code base and didn't find anything. IMO we are ok to change the interface.

In general, IMO it's a bad pattern to return interfaces from constructors instead of the actual implementation. Any such returned interface can never be extended again wihch means that the feature set of those objects is fixed forever. I also know that it is not considered great to return private struct in Go land, neither accepted by golint tools nor liked by reviewers. But this is really another strong case that this position is very questionable.

Reply:
Another argument why this change is not super harmful:

Implementors of the old interface will be warned if they use their implementation in places where the interface is used. I expect that the chance that somebody does a type check against some implementation and that suddenly fails now is super low.

Reply:
https://sourcegraph.com/search?q=-file:%22record/event.go%22+func+%5C%28%5B%5E%29%5D*%5C%29+StartLogging%5C%28.*+watch%5C.Interface&patternType=regexp&case=yes

Same result. Nobody seems to implement this, at least in OpenSource code. That's enough due diligence IMO.